File: 0030/Solution.java
public class Solution {
    public List<Integer> findSubstring(String s, String[] words) {
        LinkedList<Integer> r = new LinkedList<Integer>();
        if(words.length == 0) {
            return r;
        }
        
        int wLen = words[0].length(), len = words.length*wLen, cnt = 0, _cnt;
        if(s.length() < len) {
            return r;
        }
        
        int[] word = new int[words.length], _word = new int[words.length];
        HashMap<String, Integer> dic = new HashMap<String, Integer>();
        LinkedList<Integer> queue = new LinkedList<Integer>();
        for(int i = 0, j = 0; i < words.length; i++) {
            if(dic.containsKey(words[i])) {
                word[dic.get(words[i])]++;
            } else {
                word[j]++;
                dic.put(words[i], j);
                j++;
                cnt++;
            }
        }
        
        for(int i = 0; i < wLen; i++) {
            for(int j = 0; j < word.length; j++) {
                _word[j] = word[j];
            }
            _cnt = cnt;
            queue.clear();
            for(int j = i; j+wLen <= s.length(); j += wLen) {
                if(queue.size() == words.length) {
                    int idx = queue.removeFirst();
                    _word[idx]++;
                    if(_word[idx] == 1) {
                        _cnt++;
                    } else if(_word[idx] == 0) {
                        _cnt--;
                    }
                }
                String t = s.substring(j, j+wLen);
                Integer idx = dic.get(t);
                if(idx == null) {
                    if(j+len+wLen > s.length()) {
                        break;
                    }
                    _cnt = cnt;
                    for(int k = 0; k < word.length; k++) {
                        _word[k] = word[k];
                    }
                    queue.clear();
                } else {
                    _word[idx]--;
                    queue.add(idx);
                    if(_word[idx] == 0) {
                        _cnt--;
                    } else if(_word[idx] == -1) {
                        _cnt++;
                    }
                    if(_cnt == 0) {
                        r.add(j+wLen-len);
                    }
                }
            }
        }
        
        return r;
    }
}

File: 0030/Solution.cpp
#include <iostream>
#include <string>
#include <vector>
#include <unordered_map>
#include <unordered_set>
#include <queue>

using namespace std;

class Solution {
private:
    class helper{
    public:
        unordered_map<string, int> map, *target;
        queue<string> q;
        int max;
        int done_sz;
        helper(int max, unordered_map<string, int> *target) {
            this->max = max;
            this->target = target;
            this->done_sz = 0;
        }
        void pop(void) {
            string &s = q.front();
            map[s] --;
            if (map[s] == (*target)[s])
                done_sz ++;
            else if (map[s] == (*target)[s] - 1)
                done_sz --;
            q.pop();
        }
        void push(string &s) {
            if (q.size() == max)
                pop();
            map[s] ++;
            if (map[s] == (*target)[s])
                done_sz ++;
            else if (map[s] == (*target)[s] + 1)
                done_sz --;
            q.push(s);
        }
        int get_cnt(string &s) {
            return map[s];
        }
        bool done(void) {
            return done_sz == target->size();
        }
    };

public:
    vector<int> findSubstring(string s, vector<string>& words) {
        vector<int> ss(s.size(), -1);
        queue<helper *> slide_window;
        unordered_map<string, int> word_cnt;
        int word_sz = words.size() ? words[0].size() : 0;
        vector<int> ans;

        if (!s.size() || !words.size())
            return ans;

        for (auto &word : words)
            word_cnt[word] ++;

        for (int i = 0; i < word_sz; i ++)
            slide_window.push(nullptr);

        for (int i = 0; i < words.size(); i++) {
            int start = 0, idx;
            while ((idx = s.find(words[i], start)) != string::npos) {
                ss[idx] = i;
                start = idx + 1;
            }
        }

        for (int i = 0; i < ss.size(); i ++) {
            helper *prev = slide_window.front();
            slide_window.pop();

            if (ss[i] == -1) {
                if (prev)
                    delete prev;
                slide_window.push(nullptr);
                continue;
            }

            string &word = words[ss[i]];
            if (!prev)
                prev = new helper(words.size(), &word_cnt);
            prev->push(word);
            if (prev->done())
                ans.push_back(i - (words.size()-1) * word_sz);
            slide_window.push(prev);
        }

        return ans;
    }
};

int main(int argc, char *argv[]) {
    Solution solution;
    vector<string> words;
    vector<int> ans;
    string s(argv[1]);

    for (int i = 2; i < argc; i ++) {
        string word(argv[i]);
        words.push_back(word);
    }

    ans = solution.findSubstring(s, words);

    for (auto i : ans)
        cout << i << " ";
    cout << endl;

    return 0;
}
